tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15358,10 +15358,52 @@ void DiagnoseNodeEntry(Sema &S, FunctionDecl *FD, HLSLShaderAttr *Attr) {
   return;
 }
 
+void TryAddShaderAttrFromTargetProfile(Sema &S, FunctionDecl *FD) {
+  const std::string &EntryPointName = S.getLangOpts().HLSLEntryFunction;
+
+  // if this is the Entry FD, then add the target profile
+  // shader attribute to the FD and carry on with validation
+  // otherwise, with no shader attribute, just return
+  if (EntryPointName.empty()) {
+    return;
+  }
+  
+  // There's no way we should be adding a shader attribute
+  // to a Function Decl if it doesn't even have an identifer
+  if (!FD->getIdentifier()) {
+    return;
+  }
+
+  // if this FD isn't the entry point, then there's no
+  // shader attribute to work with, so just return
+  if (EntryPointName != FD->getIdentifier()->getName() || 
+    (EntryPointName == FD->getIdentifier()->getName() && !FD->isGlobal())) {
```

Comment:
I don't think patch constant functions can have attributes even though they are kind of entry points. I wanted to float the issue in case others know better. If they can, this code will reject them and that's a problem.